src/CGIHandler.cpp

```diff
@@ -0,0 +1,352 @@
+ *
+ * @param bufferArgv A reference to a vector of strings that is used
+ *                   to populate the argv vector.
+ * @param argv A reference to a vector of char pointers that is
+ *             populated with pointers to the strings in bufferArgv.
+ */
+void CGIHandler::setArgv(std::vector<std::string>& bufferArgv, std::vector<char*>& argv)
+{
+	bufferArgv.push_back(m_cgiPath);
+	bufferArgv.push_back(m_env["SCRIPT_FILENAME"]);
+	for (std::vector<std::string>::iterator iter = bufferArgv.begin(); iter != bufferArgv.end(); iter++)
+		argv.push_back(&(*iter).at(0));
+	argv.push_back(NULL);
+}
+
+// HELPER FUNCTIONS
+
+/**
+ * @brief Extracts the path information from a given path based on the CGI extension.
+ *
+ * This function searches for the CGI extension within the provided path. If the extension
+ * is found, it then looks for the first occurrence of a '/' character after the extension.
+ * The substring starting from this '/' character is considered the path information and is returned.
+ * If the CGI extension or the '/' character is not found, an empty string is returned.
+ *
+ * @param path The full path from which to extract the path information.
+ * @return A string containing the path information if found, otherwise an empty string.
+ */
+std::string CGIHandler::extractPathInfo(const std::string& path)
+{
+	size_t extensionStart = path.find(m_cgiExt);
+	if (extensionStart == std::string::npos)
+		return "";
+	std::string tmp = path.substr(extensionStart);
+	size_t pathInfoStart = tmp.find('/');
+	if (pathInfoStart == std::string::npos)
+		return "";
+	return tmp.substr(pathInfoStart);
+}
+
+/**
+ * @brief Extracts the script path from the given path based on the CGI extension.
+ *
+ * This function searches for the CGI extension within the provided path and
+ * returns the substring from the beginning of the path up to and including the
+ * CGI extension. If the CGI extension is not found, an empty string is returned.
+ *
+ * @param path The full path from which to extract the script path.
+ * @return std::string The extracted script path or an empty string if the CGI extension is not found.
+ */
+std::string CGIHandler::extractScriptPath(const std::string& path)
+{
+	size_t extensionStart = path.find(m_cgiExt);
+	if (extensionStart == std::string::npos)
+		return "";
+	return path.substr(0, extensionStart + m_cgiExt.size());
+}
+
+/**
+ * @brief Registers signal handlers for various signals.
+ *
+ * This function sets the default signal handlers for SIGINT, SIGTERM, and SIGQUIT,
+ * and ignores the SIGHUP signal. If any of the signal registrations fail, an error
+ * message is logged and the function returns false.
+ *
+ * @return true if all signal handlers are successfully registered, false otherwise.
+ */
+bool registerChildSignals()
+{
+	// NOLINTNEXTLINE(cppcoreguidelines-pro-type-cstyle-cast, performance-no-int-to-ptr)
+	if (std::signal(SIGINT, SIG_DFL) == SIG_ERR) {
```

Comment:
I thought the default behaviour would be to terminate the process at the point where the signal was received? Therefore we should not reach execve?

Yeah, in theory we would need a signal handler which closes all pipe fds and then exits. But I don't know how to pass the fds to such a handler. Then again how likely is it to send a signal exactly when the fds are duplicated? The whole thing is over before a manual signal can even be transmitted via the keyboard.

Reply:
I thought we set the signals for after when we execve'd!

In that case I think it's smarter to not do anything: since it only effects the small frame between succesful fork() and execve() we introduce a place where it can additionally error (resetting signal handlers) for almost zero upside.

Since dup2 also errors with EINTR on signal interrupt the possibility of receiving a signal which just gets sent to the signal handler is reduced even more.

Reply:
For when the new script process is running? Depending on the script and its execution time, this could be an interesting point.

So to my understanding from the man [page,](https://man7.org/linux/man-pages/man2/execve.2.html) if we don't register signals beforehand in the child, the new process created by execve will nonetheless handle signals, just with their default behaviour. Therefore it can be interrupted and everything.

I don't quite understand your point with the EINTR error code. But we agree on it not making much sense to handle signals for the small time fraction of dup2 and chdir, so we remove the signals completely out of the child (?)

Reply:
Yes we agree. But I think we could / should ignore SIGQUIT -> see other convi.

Trying to explain what I meant with EINTR (you can resolve convi): 

The forked process will inherit the installed signal handler. Our signal handler doesn't do much, it just updates a global atomic int. In server / main process we use the signal information to break out of the event loop.
But: since the signal handler is also active in forked process we catch the signal, and it doesn't affect the process. After execve() the signal handler gets reset to default.

This means the code, which is executed between succesful fork() and execve() could "ignore" a signal (e.g. a SIGINT) -> actually it would update the atomic int. But this effectively ingores the signal, since we don't check this value in the forked process.

However in this part of the code we mainly use C functions dup2() and chdir(). If the dup2() is interrupted by a signal it returns an error, with errno set to EINTR. This makes the surface arrea for a ignored signal even smaller.

Reply:
Ok, now I get what you mean. Basically we would not react to a signal only in chdir() which is highly unlikely to hit.

I have removed all the signals except for SIGQUIT which is set to be ignored now. That will be propagated to the new process transformed into by execve.